plugins/nf-float/src/main/com/memverge/nextflow/FloatGridExecutor.groovy

```diff
@@ -60,88 +64,139 @@ class FloatGridExecutor extends AbstractGridExecutor {
+    }
+
+    private List<String> getCmdPrefixForJob(String jobID) {
+        final oc = floatJobs.getOc(jobID)
+        return floatConf.getCliPrefix(oc)
+    }
+
+    /**
+     * Retrieve the prefix of the submit command.
+     * use round robin for multiple op-center
+     *
+     * @return
+     */
+    private List<String> getSubmitCmdPrefix() {
+        final i = serial.incrementAndGet()
+        final addresses = floatConf.addresses
+        final address = addresses[i % (addresses.size())]
+        return floatConf.getCliPrefix(address)
+    }
+
+    String toCmdString(List<String> floatCmd) {
+        def ret = floatCmd.join(" ")
+        final toReplace = [
+                ("-p " + floatConf.password): "-p ***",
+                (floatConf.s3accessKey)     : "***",
+                (floatConf.s3secretKey)     : "***",
+        ]
+        for (def entry : toReplace.entrySet()) {
+            if (!entry.key) {
+                continue
             }
+            ret = ret.replace(entry.key, entry.value)
         }
-        if (!config.mem && !config.memory) {
-            log.info '[float] missing "memory": size of memory in GB, use default'
-        }
+        return ret
     }
 
-    private String getNfs(TaskRun task) {
-        def conf = task.config
-        String nfs
-        if (!conf.nfs) {
-            nfs = floatConf.nfs
-        } else {
-            nfs = conf.nfs
-        }
-        int proto = nfs.indexOf('//')
-        int colon = nfs.indexOf(':', proto + 2)
-        if (colon == -1) {
-            // use the workDir as local mount point
-            nfs += ':' + session.workDir
-        }
-        return nfs
+    private static def warnDeprecated(String deprecated, String replacement) {
+        log.warn "[flaot] process `$deprecated` " +
+                "is no longer supported, " +
+                "use $replacement instead"
     }
 
-    private String getCpu(TaskRun task) {
-        def cpu = task.config.cpu
-        if (cpu) {
-            return cpu.toString()
+    private static def validate(TaskRun task) {
+        if (task.config.nfs) {
+            warnDeprecated('nfs', '`float.nfs` config option')
+        }
+        if (task.config.cpu) {
+            warnDeprecated('cpu', '`cpus` directive')
         }
-        cpu = task.config.cpus
-        if (cpu) {
-            return task.config.getCpus().toString()
+        if (task.config.mem) {
+            warnDeprecated('mem', '`memory` directive')
+        }
+        if (task.config.image) {
+            warnDeprecated('image', '`container` directive')
         }
-        return floatConf.cpu
     }
 
-    String getImage(TaskRun task) {
-        def image = getImageFromConf(task)
+    String getContainer(TaskRun task) {
+        def image = task.getContainer()
+        if (!image) {
+            image = DFT_IMAGE
+            log.warn "container image not specified for" +
+                    "${task.id}, use default image: $DFT_IMAGE"
+        }
```

Comment:
The default image is helpful for some demo.  You know, sometimes we what to demo some simple `hello world`, loop print or naive scatter.  Most of the time, a minimal fedora or ubuntu will do the trick.  It would save them the effort of inputing the full `"quay.io/fedora/fedora-minimal"`.

To make things worse, when user just input `fedora/fedora-minimal` without specify the default registry, MMC will report error because it by default use the docker hub registry and cannot find this image.

So I would prefer to leave the default image here.  It would fail later when the image doesn't have the necessary binaries.

Reply:
The reason your default image doesn't work with the custom registry is exactly because you are bypassing `task.getContainer()`, which applies the custom registry for you. This is why I think it would be better to not have a default at all, and instead tell users to add something like this to their config:

```groovy
process.container = 'fedora/fedora-minimal'
```

It will behave as a fallback value for any process that doesn't define it's own container, and Nextflow will apply any custom registry for you.

It's still fine whatever you guys want to do, I'm just telling you what the standard practice is.

Reply:
OK.  I get it now.  You recommend to use NextFlow's default container and registry to manage that.  I agree that we should follow what NextFlow's practice is.  Then I think a simple get container will do.  Let me refine this.

Reply:
My original thought is that at least the plugin has a backup plan for the default image.  But I agree that align to the best practice for within the whole solution is more important.

Reply:
Oh, yes.  That reminds me.  I should add process.container in the read me.  Thanks.